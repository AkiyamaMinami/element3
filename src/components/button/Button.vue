<script setup lang="ts">
// import { props } from './props'
import { computed, inject, toRefs, withDefaults } from 'vue'
import type { ComputedRef, Ref } from 'vue'
import { useGlobalOptions } from '../../hooks/globalConfig'
const props = withDefaults(defineProps<ButtonProps>(), {
  nativeType: 'button',
})

type ButtonType = 'primary'
| 'success'
| 'warning'
| 'danger'
| 'info'
| 'text'

type ButtonSize = 'large' | 'medium' | 'small' | 'mini'

type ButtonNativeType = 'button' | 'submit' | 'reset' | 'menu'

interface ButtonProps {
  size?: ButtonSize
  type?: ButtonType
  nativeType?: ButtonNativeType
  plain?: boolean
  round?: boolean
  circle?: boolean
  loading?: boolean
  disabled?: boolean
  icon?: string
}

const useButtonSize = function (size: Ref<ButtonSize>): ComputedRef<ButtonSize> {
  const globalConfig = useGlobalOptions()
  return computed(() => {
    // const elFormItem = inject('elFormItem', null)
    // || elFormItem?.elFormItemSize
    return size?.value || globalConfig.size
  })
}

const useClasses = function ({ props, size, disabled }) {
  return computed(() => {
    return [
      size.value ? `el-button--${size.value}` : '',
      props.type ? `el-button--${props.type}` : '',
      {
        'is-plain': props.plain,
        'is-round': props.round,
        'is-circle': props.circle,
        'is-loading': props.loading,
        'is-disabled': disabled.value,
      },
    ]
  })
}

const { size, disabled } = toRefs(props)
const buttonSize = useButtonSize(size)
// const buttonDisabled = useButtonDisabled(disabled)
const classes = useClasses({
  props,
  disabled,
  size: buttonSize,
})

// const useButtonDisabled = (disabled: Ref) => {
//   return computed(() => {
//     const elForm = inject('elForm', null)

//     return disabled?.value || elForm?.disabled
//   })
// }
</script>

<template>
  <button
    class="el-button"
    :class="classes"
    :type="nativeType"
    :disabled="disabled || loading"
  >
    <i v-if="loading" class="el-icon-loading" data-testid="loadingIcon" />
    <i v-else-if="icon" :class="icon" data-testid="icon" />
    <span v-if="$slots.default">
      <slot />
    </span>
  </button>
</template>

<style lang="scss">
@import '../../theme/src/button.scss'
</style>>

