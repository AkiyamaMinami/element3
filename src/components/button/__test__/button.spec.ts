import { describe } from 'vitest'
import { render } from '@testing-library/vue'
import Button from '../Button.vue'

describe('Button.vue', () => {
  it('用户可以通过 slot 的方式，定义组件的内容', () => {
    const { getByText } = render(Button, {
      slots: {
        default: '默认按钮',
      },
    })
    expect(getByText('默认按钮')).toBeInTheDocument()
  })

  // primary / success / warning / danger / info / text
  it('type 渲染不同类型按钮，正常、警告、错误', () => {
    const { getByRole } = render(Button, {
      props: {
        type: 'primary',
      },
    })
    expect(getByRole('button')).toHaveClass('el-button--primary')
  })

  // 'medium' | 'small' | 'mini';
  it('根据size 渲染按钮大小', () => {
    const { getByRole } = render(Button, {
      props: {
        size: 'medium',
      },
    })
    expect(getByRole('button')).toHaveClass('el-button--medium')
  })

  // 根据round设置不用圆角
  it('根据 是否存在 round 渲染按钮大小', () => {
    const { getByRole } = render(Button, {
      props: {
        round: true,
      },
    })
    expect(getByRole('button')).toHaveClass('is-round')
  })
})
