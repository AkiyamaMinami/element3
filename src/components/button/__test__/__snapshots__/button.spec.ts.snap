// Vitest Snapshot v1

exports[`Button.vue > snapshot 1`] = `
<body>
  <div>
    <button
      class="el-button"
      type="button"
    >
      <!--v-if-->
      <!--v-if-->
    </button>
  </div>
</body>
`;
